### ISSUE TITLE
feat: class wrappers for functional-only operators (mask API symmetry)

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/finitevolX#206]
## Summary

Give the functional-only diagnostics, the Arakawa Jacobian and the spherical geostrophic velocity a **class form** that participates in the operator-attribute mask API, so `mask=` works the same way everywhere.

Today every class operator takes its mask at construction and zeroes dry cells of its output stagger automatically:

```python
vort = fvx.Vorticity2D(grid=grid, mask=mask)
zeta = vort.relative_vorticity(u, v)          # dry X-corners are exactly 0
```

but the diagnostics only exist as mask-free functions, so users must know the output stagger and the right mask field themselves:

```python
ke = fvx.kinetic_energy(u, v) * mask.h                        # T-points
ss = fvx.shear_strain(u, v, dx, dy) * mask.xy_corner_strict   # X-points
J  = fvx.arakawa_jacobian(f, g, dx, dy)                       # interior shape only — no ghost ring, can't be masked as-is
```

The functional forms **stay** (they are the mask-free Layer 2 that jit'd kernels compose). This issue adds the Layer 3 wrappers on top of them.

> This issue was rewritten on 2026-09-24 against `main` at `dd0b155`. The original text predates the operator-attribute mask design (#209, landed in #211) and referenced APIs that no longer exist; everything below has been checked against the current code.

## What changed since this issue was first written

| Then (April 2026) | Now (`main`) | Consequence for this issue |
|---|---|---|
| Masks were per-call kwargs: `op(u, v, mask=...)` (PR #204, never merged) | Masks are **class attributes** set at construction: `Op(grid=grid, mask=mask)` (#209 / #211) | New classes take `mask` in `__init__`, not in `__call__` |
| `ArakawaCGridMask`, field `mask.psi` | `Mask2D` / `Mask3D`; corner field is `mask.xy_corner_strict` | Names in the tables below |
| `tests/test_postcompute_masks.py` | Per-operator `tests/test_<op>_masks.py` + goldens registered in `tests/fixtures/_gen_golden.py` | Test plan below |
| `strain_magnitude_squared(u, v, dx, dy)`, `okubo_weiss(u, v, dx, dy)` | Both are now **pointwise**: `strain_magnitude_squared(sn, ss)`, `okubo_weiss(sn, ss, omega)`, inputs must be co-located | The class has to do the X→T interpolation the functional forms no longer do |
| `tensor_strain` returned "dilatation at T + shear at X" | Returns a single T-point array | One output stagger, `mask.h` |
| `relative_vorticity_cgrid`, `sw_potential_vorticity`, `sw_potential_vorticity_multilayer` needed wrappers | Already covered by `Vorticity2D.relative_vorticity`, `Vorticity2D.potential_vorticity` and `multilayer(...)` — **bit-identical** (max abs diff 0.0 on the canonical fixture) | Dropped from scope |
| `vertical_velocity` | Still returns `[Nz+1, Ny, Nx]`; #210 is open to change that to `[Nz, Ny, Nx]` so it can be masked by `Mask3D.w` | Blocked by #210, out of scope here |

## Where every functional operator stands today

| Function (`finitevolx/_src/operators/…`) | Output stagger | Already covered by a class? | Proposed home |
|---|---|---|---|
| `kinetic_energy(u, v)` | T | no | `Energetics2D.kinetic_energy` |
| `bernoulli_potential(h, u, v, gravity)` | T | no | `Energetics2D.bernoulli_potential` |
| `available_potential_energy(h, H, g_prime)` | T | no | `Energetics2D.available_potential_energy` |
| `shear_strain(u, v, dx, dy)` | X | no | `Strain2D.shear` |
| `tensor_strain(u, v, dx, dy)` | T | no | `Strain2D.tensor` |
| `strain_magnitude_squared(sn, ss)` | pointwise | no | `Strain2D.magnitude_squared(u, v)` — at T, after X→T of the shear |
| `okubo_weiss(sn, ss, omega)` | pointwise | no | `Strain2D.okubo_weiss(u, v)` — at T, after X→T of shear and vorticity |
| `enstrophy(omega)` | pointwise | no | new method `Vorticity2D.enstrophy(u, v)` at X |
| `potential_enstrophy(q, h)` | pointwise | no | new method `Vorticity2D.potential_enstrophy(u, v, h, f)` at X |
| `qg_potential_vorticity(psi, f0, beta, dx, dy, y, y0)` | T | no | `QGPotentialVorticity2D.__call__` |
| `stretching_term(A, psi)` | T, `[nl, Ny, Nx]` | no | `QGPotentialVorticity2D.stretching` |
| `potential_vorticity_multilayer(psi, A, …)` | T, `[nl, Ny, Nx]` | no (= `qg_pv` per layer − `stretching_term`) | `QGPotentialVorticity2D.multilayer` |
| `arakawa_jacobian(f, g, dx, dy)` | T, **interior shape** | no | `ArakawaJacobian2D.__call__`, padded to full shape |
| `geostrophic_velocity_sphere(h, f, grid, gravity)` | (U, V) | Cartesian case: `Difference2D.grad_perp`; spherical: no | new method `SphericalDifference2D.geostrophic_velocity(h, f, gravity)` |
| `relative_vorticity_cgrid`, `sw_potential_vorticity`, `sw_potential_vorticity_multilayer` | X | **yes** — `Vorticity2D` (+ `multilayer`) | nothing to do |
| `potential_vorticity(omega, f, h)` | pointwise | yes — `Vorticity2D.potential_vorticity` does the full pipeline | nothing to do |
| `total_energy`, `total_enstrophy` | scalar | — | not field operators; leave functional |
| `vertical_velocity` | `[Nz+1, …]` | — | wait for #210 |
| `coriolis_param`, `beta_param`, `coriolis_fn`, `streamfn_to_ssh`, `ssh_to_streamfn` | pointwise / scalar | — | not stencil operators; leave functional |

Grouping into a few classes with several methods each (`Energetics2D`, `Strain2D`, `QGPotentialVorticity2D`) mirrors how `Vorticity2D` and `Difference2D` are already organised, and avoids a dozen one-method classes.

### Constructor and mask wiring

Every new class follows the existing `Vorticity2D` shape exactly:

```python
class Strain2D(eqx.Module):
    grid: CartesianGrid2D
    mask: Mask2D | None
    interp: Interpolation2D

    def __init__(self, grid: CartesianGrid2D, mask: Mask2D | None = None) -> None:
        self.grid = grid
        self.mask = mask
        self.interp = Interpolation2D(grid=grid, mask=mask)   # pass-down (Pattern 2)
```

`__call__` / methods take **fields only** — no `mask` argument, no `dx`/`dy` (they come from `self.grid`).

### Mask application per method

The rule from `docs/masks.md` is "one output → one mask field, chosen by the output stagger":

| Output stagger | Mask field | Methods |
|---|---|---|
| T | `mask.h` | `Energetics2D.*`, `Strain2D.tensor`, `Strain2D.magnitude_squared`, `Strain2D.okubo_weiss`, `QGPotentialVorticity2D.*`, `ArakawaJacobian2D` |
| X | `mask.xy_corner_strict` | `Strain2D.shear`, `Vorticity2D.enstrophy`, `Vorticity2D.potential_enstrophy` |
| U / V | `mask.u` / `mask.v` | `SphericalDifference2D.geostrophic_velocity` |

Two patterns cover everything:

- **Pattern 1 — delegate, then multiply.** Where the functional form does the whole stencil (`kinetic_energy`, `bernoulli_potential`, `available_potential_energy`, `shear_strain`, `tensor_strain`, `qg_potential_vorticity`, `stretching_term`, `arakawa_jacobian`, `geostrophic_velocity_sphere`):

  ```python
  def kinetic_energy(self, u, v):
      # ke[j, i] = 1/2 * (u2_on_T[j, i] + v2_on_T[j, i])   (functional form does the U->T / V->T averages)
      out = kinetic_energy(u, v)
      if self.mask is not None:
          out = out * self.mask.h
      return out
  ```

  With `mask=None` this is **bit-identical** to the functional form — that is what the "unmasked golden" test pins.

- **Pattern 2 — pass-down through `Interpolation2D`.** Where the class has to co-locate fields before a pointwise functional form (`magnitude_squared`, `okubo_weiss`, `enstrophy`, `potential_enstrophy`):

  ```python
  def okubo_weiss(self, u, v):
      sn = self.tensor(u, v)                          # T-points
      # ss_on_T[j, i] = 1/4 * (ss[j-1/2, i-1/2] + ss[j-1/2, i+1/2] + ss[j+1/2, i-1/2] + ss[j+1/2, i+1/2])
      ss_on_T = self.interp.X_to_T(self.shear(u, v))
      omega_on_T = self.interp.X_to_T(relative_vorticity_cgrid(u, v, self.grid.dx, self.grid.dy))
      out = okubo_weiss(sn, ss_on_T, omega_on_T)      # pointwise at T
      if self.mask is not None:
          out = out * self.mask.h
      return out
  ```

  `self.interp` already carries the mask, so the intermediate X-fields are zero at dry corners before the average (same reasoning as `Vorticity2D`).

### Things that need care

1. **`ArakawaJacobian2D` must pad, and not with `interior()` for multilayer input.** `arakawa_jacobian` returns the interior `(..., Ny-2, Nx-2)`. The class restores the full `[Ny, Nx]` shape with a zero ghost ring so the output composes with the rest of the suite and can be masked. `interior(values, like)` slices **every** axis of `like`, so on a `[nl, Ny, Nx]` input it silently returns all zeros (checked). Pad the last two axes explicitly:

   ```python
   # J[..., j, i] for 1 <= j <= Ny-2, 1 <= i <= Nx-2; zero ghost ring
   out = jnp.zeros_like(f).at[..., 1:-1, 1:-1].set(arakawa_jacobian(f, g, dx, dy))
   ```

2. **Division under a mask → `jnp.where`, not `*`.** `potential_enstrophy` divides by `h_on_q`, and `geostrophic_velocity_sphere` divides by `f` and `cos φ`. Under pass-down masking the denominators are zero at dry points, giving NaN; follow `Vorticity2D.potential_vorticity` and finish with `jnp.where(self.mask.<field>, out, 0.0)` so dry cells are exactly zero and wet NaNs (a real bug signal) survive.

3. **`Vorticity2D.potential_enstrophy(u, v, h, f)`** should reuse `self.potential_vorticity(u, v, h, f)` and `h_on_q = self.interp.T_to_X(h)`, then call the pointwise `potential_enstrophy(q, h_on_q)`; `enstrophy(u, v)` is `enstrophy(self.relative_vorticity(u, v))`. Both stay at X and use `mask.xy_corner_strict`.

4. **`QGPotentialVorticity2D`** holds nothing but `grid` and `mask`; `__call__(psi, f0, beta, y, y0)` mirrors the functional signature minus `dx, dy`. `stretching(A, psi)` and `multilayer(psi, A, f0, beta, y, y0)` (= per-layer PV via `fvx.multilayer(self)` minus `stretching`) keep the `[nl, Ny, Nx]` layout. Same padding caveat as (1) for the multilayer paths.

5. **`SphericalDifference2D.geostrophic_velocity`** is a method, not a new class: `SphericalDifference2D` already owns a `SphericalGrid2D` and a `Mask2D | None`, and `Difference2D.grad_perp` is the Cartesian precedent for "streamfunction/height → (u, v) on faces". Output `(u_g * mask.u, v_g * mask.v)` via `jnp.where` per (2).

6. **Naming.** Existing classes carry the `2D` suffix (`Vorticity2D`, `Coriolis2D`); the new ones do too. The original proposal's suffix-less `KineticEnergy(grid)` is dropped.

7. **3-D.** None of these have 3-D functional forms today; `Vorticity3D` is the only place a 3-D method would be added (`enstrophy`), and only if it's trivial. Multilayer use goes through `fvx.multilayer(op.method)` as documented in `docs/multilayer_vs_3d.md`.

## Tests

Mirror `tests/test_coriolis_masks.py`. For every new method, one `tests/test_diagnostics_masks.py` (plus `tests/test_jacobian_masks.py`, and the spherical method in `tests/test_spherical_masks.py`) with the four standard checks on the canonical 18×18 cross-shaped-island fixture from `tests/fixtures/inputs.py`:

1. **unmasked golden** — `Op(grid)(...)` matches the committed `.npz` (and, in a separate assertion, matches the functional form bit-for-bit);
2. **masked golden** — `Op(grid, mask=make_mask_2d())(...)` matches the committed `.npz`;
3. **all-ocean invariant** — `mask=make_mask_2d_all_ocean()` gives exactly the unmasked output;
4. **dry-cell zero** — every dry cell of the output stagger (`~mask.h`, `~mask.xy_corner_strict`, `~mask.u`, `~mask.v`) is exactly `0`, and the output is finite everywhere.

Goldens: add a `_diagnostics_entries(...)` (and a jacobian entry, and the spherical method under `_spherical_entries`) to `_register_all` in `tests/fixtures/_gen_golden.py`, run `uv run python tests/fixtures/_gen_golden.py`, and commit the new `tests/fixtures/golden/<Op>__<method>__{unmasked,masked}.npz` files. Multi-output methods (`geostrophic_velocity`) are saved as `out0`, `out1` automatically.

The existing `tests/test_diagnostics.py::TestMaskSupport` pins the call-site convention for the functional forms; keep it.

## Docs

- `docs/masks.md`, *Operator API*: add the new classes/methods to the Layer 3 row of the three-layer table and to the stagger → mask field table; replace the `ke = kinetic_energy(u, v) * mask.h` call-site example with the class form and keep the call-site form as the Layer 2 alternative.
- `docs/api/diagnostics.md`: add the class entries next to the functional ones (`::: finitevolx.Energetics2D`, …); `docs/api/vorticity.md`: `ArakawaJacobian2D` and the two new `Vorticity2D` methods. The spherical operators have no `docs/api/` page today, so `geostrophic_velocity` only needs its docstring here (a spherical API page is a separate task).
- `docs/spatial_operators.md` operator table: new rows.

## Acceptance criteria

- [ ] `Energetics2D(grid, mask=None)` with `kinetic_energy`, `bernoulli_potential`, `available_potential_energy` (T, `mask.h`)
- [ ] `Strain2D(grid, mask=None)` with `shear` (X), `tensor` (T), `magnitude_squared` (T), `okubo_weiss` (T)
- [ ] `Vorticity2D.enstrophy(u, v)` and `Vorticity2D.potential_enstrophy(u, v, h, f)` (X, `mask.xy_corner_strict`, `jnp.where` for the division)
- [ ] `QGPotentialVorticity2D(grid, mask=None)` with `__call__`, `stretching`, `multilayer` (T, `mask.h`)
- [ ] `ArakawaJacobian2D(grid, mask=None)` returning full `[..., Ny, Nx]` with a zero ghost ring, padded with `.at[..., 1:-1, 1:-1]`
- [ ] `SphericalDifference2D.geostrophic_velocity(h, f, gravity)` → `(u_g, v_g)` masked by `mask.u` / `mask.v`
- [ ] With `mask=None`, every class method is bit-identical to the corresponding functional form (asserted in tests)
- [ ] Four-check mask regression tests per method, goldens committed, `_gen_golden.py` registrations added
- [ ] New classes exported from `finitevolx/__init__.py` (and listed in `__all__`)
- [ ] Docs updated as above
- [ ] `uv run pytest tests -v`, `ruff check .`, `ruff format --check .`, `ty check finitevolx` all clean

## Decisions to confirm before starting

1. **Grouped classes vs. one class per function.** This rewrite proposes `Energetics2D` / `Strain2D` / `QGPotentialVorticity2D` (matching `Vorticity2D`'s several-methods style) instead of the original one-class-per-function list. Say if you prefer the flat list.
2. **Pointwise helpers with no natural stagger** (`enstrophy`, `potential_enstrophy`) are folded into `Vorticity2D`, which computes ζ itself, rather than given a `stagger=` switch. `strain_magnitude_squared` / `okubo_weiss` are pinned to **T-points** (shear and vorticity interpolated X→T). If you want X-point variants too, that's an extra method each.
3. **Spherical geostrophic velocity as a `SphericalDifference2D` method** rather than a standalone class.

## Out of scope

- Removing or changing the functional forms — they stay as the mask-free Layer 2.
- `total_energy`, `total_enstrophy`, `coriolis_*`, `streamfn_to_ssh`, `ssh_to_streamfn` — scalars / pointwise conversions, not stencil operators.
- `vertical_velocity` — blocked by #210 (shape change so `Mask3D.w` applies); do it there.
- Adaptive-stencil dispatch at coastlines for these operators — #205.
- The `pv_flux_*` docstring clarification — #203.
- somax migration to the class forms — track in somax if wanted.
